content/chainguard/chainguard-images/getting-started/nginx/index.md

```diff
@@ -0,0 +1,262 @@
+
+Copy this code to your `index.html` file, save, and close it.
+
+Next, create a CSS file named `stylesheet.css` using the text editor of your choice. We'll use `nano` to demonstrate.
+
+```shell
+nano stylesheet.css
+```
+
+Copy the following code to your `stylesheet.css` file. 
+
+```CSS
+/*
+Chainguard Academy nginx Demo Application
+*/
+
+body {
+    text-align: center;
+    background-color: #fcebfc;
+    font-family: Arial, sans-serif;
+}
+
+h1 {
+    color: #df45e6;
+}
+
+h2, h3 {
+    color: #9745e6;
+}
+
+p {
+    color: #583ce8
+}
+
+.corners {
+    border-radius: 10px;
+}
+
+```
+
+After copying the code into the `stylesheet.css` file, save and close it.
+
+Next, you will pull down the `inky.png` file using `curl`. Always [inspect the URL](https://raw.githubusercontent.com/chainguard-dev/edu-images-demos/734e6171ee69f0e0dbac95e6ebc1759ac18bf00a/nginx/data/inky.png) before downloading it to ensure it comes from a safe and trustworthy location.
+
+```shell
+curl -O https://raw.githubusercontent.com/chainguard-dev/edu-images-demos/734e6171ee69f0e0dbac95e6ebc1759ac18bf00a/nginx/data/inky.png
+```
+
+Now, return to the `nginxdemo` directory you made earlier. Here we will create the `nginx.conf` configuration file used by nginx to run the local web server. We will demonstrate this using `nano`.
+
+```shell
+nano nginx.conf
+```
+
+Copy the following code into the configuration file you created. In the `location` directive, be sure to update the configuration to reference the path from root to the `data` directory you created in a previous step.
+
+```nginx
+worker_processes  1;
+
+events {
+    worker_connections  1024;
+}
+
+http {
+    include       mime.types;
+    default_type  application/octet-stream;
+
+    sendfile        on;
+
+    keepalive_timeout  65;
+
+    server {
+
+        listen 1313;
+        server_name localhost;
+        charset koi8-r;
+
+        location / {
+            root /Users/username/nginxdemo/data; # Update the file path for your system
+
+        }
+
+    }
+
+    include servers/*;
+
+}
+```
+
+Once you are finished editing the configuration file, save and close it.
+
+Create a new file named `mime.types` using a text editor of your choice. We will use `nano` to demonstrate.
+
+```shell
+nano mime.types
+```
+
+Copy and paste the following code into your `mime.types` file. This file will allow nginx to handle the HTML, CSS, and png files we created when rendering the webserver. 
+
+```nginx
+types {
+    text/html                                        html htm shtml;
+    text/css                                         css;
+    image/png                                        png;
+}
+```
+
+Save and close the `mime.types` file when you are finished editing it.
+
+Next, copy the filepath to the `nginx.conf` file you created earlier. Replacing the example path with this copied path, execute the following command to initialize the nginx server:
+
+```shell
+nginx -c /Users/username/nginxdemo/nginx.conf
```

Comment:
This does run on Ubuntu 22, but I get a warning:

```
nginx: [emerg] open() "/usr/share/nginx/nginx.conf" failed (2: No such file or directory)
```

You might also want to warn that it has to be an absolute path, I tried relative first and it doesn't work. (I know you showed an absolute path, but guessing that a relative will also work might be somewhat common, and it's unexpected that it doesn't.)

Reply:
Did you get this warning only when you used a relative path? Or did you get it even with using an absolute path? Thanks for mentioning this.

Reply:
This is actually worse than I thought. It doesn't seem possible to override the default accessing of log files in the /var/ directory, which by default is not accessible to a user in Ubuntu.  I've tried adding directives to write access_log or error_log to directories with permissions, but it still seems to want to access the default location first. There's supposed to be an -e flag to suggest error log location, but that option is removed, seemingly.If sudo is used, then the user gets a403 forbidden when visiting locaclhost:4000.

This is the command that got it to the point of working, though with a warning:

```
sudo chown -R patrick:patrick /var/log/nginx
```

The only real solution I've found here is having the user change permissions on the folder with the logs (/var/logs/nginx). In general, I'd say the nginx command line option is pretty useless, and nginx mostly needs to be configured globally and run with systemd. The default packaging for ubuntu is also pretty silly (logging to a folder with no permissions).

Not sure how you want to handle this. You may want to put in a note in about permissions. Typically, nginx is configured globally and run from systemd or similar, that's how I've used it, only using the command line utility to reload the global config. Happy to do a meeting to work through this, though I'll be away next week. 

@ltagliaferri  in case you have thoughts. I think Michelle would need access to a linux box to figure this part out. The issue here is in the non-containerized part of the tutorial.

Reply:
When working on the demo, I intentionally removed the access_log and error_log directives in the nginx.conf file. I did this because I was running into issues with the file paths they wanted to default to, and I thought that it may be confusing for users to have to change these paths themselves, considering the default location depends on how you install nginx (I used Homebrew, for example). Taking these directives out seemed to resolve the problem and I thought that would be okay for the scope of this demo.

It sounds like you already tried to route the directives to a new location, which I think I tried (before I opted to take the directives out entirely) and I was not having success there. 

I agree that the command line option is quite useless. I would have used the default configuration file, but with the location of it varying with each installation method, I didn't think this made sense. 

:// I can give it a shot on Linux this weekend. I've been meaning to do an install on a spare drive so now is my chance!

Reply:
@smythp, I was able to test this out on Linux Mint and I reached a similar conclusion. Without changing the permissions on that default folder, I was not able to get things running. 

If I do update the permissions, and run nginx using `nginx -c /home/michelle/nginxdemo/nginx.conf`, then the local site loads fine.
However, I get this warning when I stop the server:
`nginx: [warn] the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:1`


For whatever reason, if I use `sudo` with `nginx -c /home/michelle/nginxdemo/nginx.conf` I get a 403 forbidden error when navigating to the local site. With `sudo nginx -s quit` there are no command-line warnings. 

My solution for now will be to mention this in the tutorial. I will add that nginx may need permissions to write to default log directories if it does not already have them, and exemplify this. 

I would choose to do a global configuration, however, I was worried that doing so would make the tutorial less flexible. From what I can tell, nginx is installed in different places depending on the OS and installation methods, so I wouldn't be able to cover each case. I installed using Homebrew on mac and apt on Linux and even then that's just two examples. Let me know what you think, I'll try to accomodate this permission problem in the tutorial as best I can, but if we think that things should be structured differently then we can make those changes

Reply:
I think following your solution makes sense, I would not spend more time on this.